src/libs/actions/Welcome/OnboardingFlow.ts

```diff
@@ -0,0 +1,126 @@
+        return;
+    }
+
+    const rootState = navigationRef.getRootState();
+    const adaptedState = rootState;
+    const lastRouteIndex = (adaptedState?.routes?.length ?? 0) - 1;
+    const onBoardingModalNavigatorState = adaptedState?.routes[lastRouteIndex]?.state;
+    if (!onBoardingModalNavigatorState || onBoardingModalNavigatorState?.routes?.length > 1) {
+        return;
+    }
+
+    let adaptedOnboardingModalNavigatorState = {} as Readonly<PartialState<NavigationState>>;
+    if (currentRoute?.name === SCREENS.ONBOARDING.PERSONAL_DETAILS && selectedPurpose === CONST.ONBOARDING_CHOICES.MANAGE_TEAM) {
+        adaptedOnboardingModalNavigatorState = {
+            index: 2,
+            routes: [
+                {
+                    name: SCREENS.ONBOARDING.PURPOSE,
+                    params: currentRoute?.params,
+                },
+                {
+                    name: SCREENS.ONBOARDING.WORK,
+                    params: currentRoute?.params,
+                },
+                {...currentRoute},
+            ],
+        } as Readonly<PartialState<NavigationState>>;
+    } else {
+        adaptedOnboardingModalNavigatorState = {
+            index: 1,
+            routes: [
+                {
+                    name: SCREENS.ONBOARDING.PURPOSE,
+                    params: currentRoute?.params,
+                },
+                {...currentRoute},
+            ],
+        } as Readonly<PartialState<NavigationState>>;
+    }
+
+    adaptedState.routes[lastRouteIndex].state = adaptedOnboardingModalNavigatorState;
+    navigationRef.resetRoot(adaptedState);
+}
+
+/**
+ * Start a new onboarding flow or continue from the last visited onboarding page.
+ */
+function startOnboardingFlow() {
+    const currentRoute = navigationRef.getCurrentRoute();
+    const {adaptedState} = getAdaptedStateFromPath(getOnboardingInitialPath(), linkingConfig.config, false);
+    const focusedRoute = findFocusedRoute(adaptedState as PartialState<NavigationState<RootStackParamList>>);
+    if (focusedRoute?.name === currentRoute?.name) {
+        return;
+    }
+    navigationRef.resetRoot(adaptedState);
+}
+
+function getOnboardingInitialPath(): string {
+    const state = getStateFromPath(onboardingInitialPath, linkingConfig.config);
+    if (state?.routes?.at(-1)?.name !== NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR) {
+        return `/${ROUTES.ONBOARDING_ROOT.route}`;
+    }
+
+    return onboardingInitialPath;
+}
+
+function clearInitialPath() {
+    onboardingInitialPath = '';
+}
+
+/**
+ * Onboarding flow: Go back to the previous page.
+ * Since there is no `initialRoute` for `onBoardingModalNavigator`,
+ * firstly, adjust the current onboarding modal navigator to establish the correct stack order.
+ * Then, navigate to the previous onboarding page using the usual `goBack` function.
+ */
+function goBack() {
+    adaptOnboardingRouteState();
+    Navigation.isNavigationReady().then(() => {
+        Navigation.goBack();
```

Comment:
Coming from the BZ of #50177
This condition is missing handling for the case where the user navigates back using the hardware button on Android and the onboarding route state does not initialize properly. The solution for this issue is to reset the onboarding state when setting up the router.